README.md
# Python QSYS QRC Wrapper
* Control QSC QSYS Core devices with python

# ToDo
* Flaskify... 
* Document, document, document

# Use
* For each QSC Core on the network instantiate a "Core" class
* When adding control objects they will "cast" themselves to the parent core class
    * The parent Core class instance is required as keyword arg "parent" when creating control objects
```python
#!/usr/bin/python3
import time
from qsys.classes import Core,Control,ChangeGroup

#returns epoch time
from qsys.helpers import epoch

def main():
    #See qsys.py for parameters in Core class
    #The initiail EngineStatus response parameters from the device will get added to Core.__dict__
    #You can pass "port" as well, but it defaults to 1710
    myCore = Core(Name='myCore',User='',Password='',ip='192.168.61.2')

    #Open the socket,creates "listen" and "keepalive" threads
    myCore.start()

    #ValueType can be a list of potential value types [int,float] or a single type "str" etc
    #This object is assumed to be a "gain" control object, so we can pass [int,float]
    gainControlObject = Control(parent=myCore,Name='namedControlInQsysDesigner',ValueType=[int,float])

    #To constantly monitor the state of your object use a ChangeGroup
    #You need to a ChangeGroup instance to add control objects and set polling rates
    #Parameters that are capitalize are that way because of the QRC parameter protocol
    #Id in this case is just the name of the ChangeGroup
    myChangeGroup = ChangeGroup(parent=myCore,Id='myChangeGroup')
    myChangeGroup.AddControl(gainControlObject)

    #Allow the socket time to connect and parse the initial responses
    time.sleep(2)

    #Set the change group auto poll rate
    #This rate is fast, your mileage may vary
    myChangeGroup.AutoPoll(Rate=0.1)

    #Value = value to set object to
    #TransId = QRC id parameter for transaction ID
    gainControlObject.set(Value=10,TransId=epoch())

    while True:
        print(gainControlObject.state)
        time.sleep(1)

if __name__ == '__main__':
    main()
```

# Notes
* In development, versions will change rapidly. This version doesn't do much yet.. stand by

# References
* https://q-syshelp.qsc.com/Content/External_Control/Q-Sys_Remote_Control/QRC.htm
* https://github.com/rakibtg/Python-Chat-App - simple chat app example with SocketIO, use for UI


RELEASES.md
# 0.0.4
* Fixed project structure

# 0.0.3
* Messed up project structure

# 0.0.2
* Basic functionality

# 0.0.1
* It doesn't work :D


test.py
#!/usr/bin/python3
import time
from qsys.classes import Core,Control,ChangeGroup

#returns epoch time
from qsys.helpers import epoch

def main():
    #See qsys.py for parameters in Core class
    #The initiail EngineStatus response parameters from the device will get added to Core.__dict__
    #You can pass "port" as well, but it defaults to 1710
    myCore = Core(Name='myCore',User='',Password='',ip='192.168.61.2')

    #Open the socket,creates "listen" and "keepalive" threads
    myCore.start()

    #ValueType can be a list of potential value types [int,float] or a single type "str" etc
    #This object is assumed to be a "gain" control object, so we can pass [int,float]
    gainControlObject = Control(parent=myCore,Name='namedControlInQsysDesigner',ValueType=[int,float])

    #To constantly monitor the state of your object use a ChangeGroup
    #You need to a ChangeGroup instance to add control objects and set polling rates
    #Parameters that are capitalize are that way because of the QRC parameter protocol
    #Id in this case is just the name of the ChangeGroup
    myChangeGroup = ChangeGroup(parent=myCore,Id='myChangeGroup')
    myChangeGroup.AddControl(gainControlObject)

    #Allow the socket time to connect and parse the initial responses
    time.sleep(2)

    #Set the change group auto poll rate
    #This rate is fast, your mileage may vary
    myChangeGroup.AutoPoll(Rate=0.1)

    #Value = value to set object to
    #TransId = QRC id parameter for transaction ID
    gainControlObject.set(Value=10,TransId=epoch())

    while True:
        print(gainControlObject.state)
        time.sleep(1)

if __name__ == '__main__':
    main()


qsys/classes.py
#!/usr/bin/python3
import json,socket,time,threading
from qsys.helpers import required_args,lineno,epoch
from logg3r import Log
import copy

init_logger = Log(log_path="./test_logs/",name='init_logger',level=1)

ErrorCodes = {
    -32700:'Parse error. Invalid JSON was received by the server.',
    -32600:'Invalid request. The JSON sent is not a valid Request object.',
    -32601:'Method not found.',
    -32602:'Invalid params.',
    -32603:'Server error.',
    2:'Invalid Page Request ID',
    3:'Bad Page Request - could not create the requested Page Request',
    4:'Missing file',
    5:'Change Groups exhausted',
    6:'Unknown change croup',
    7:'Unknown component name',
    8:'Unknown control',
    9:'Illegal mixer channel index',
    0:'Logon required'
}

IgnoreKeys = ['sock','logger','parent','ip','port','ValueType']

#contains all method, control, etc objects
#handles parsing, etc - instantiate one per controlled QSC core
class Core():
    def __init__(self,**kwargs):
        required = {'User':str,
                    'Password':str,
                    'Name':str,
                    'ip':str}
        self.success,error = required_args(kwargs,required)
        if(self.success):
            try: 
                self.port = kwargs['port']
            except KeyError: 
                self.port = 1710 #standard port

            self.__dict__.update(**kwargs)
            self.logger = Log(log_path="./test_logs/",name='Core_{ip}'.format(**self.__dict__),level=1)
            self.logger.log(lineno()+"<qsys.classes.Core object - {}> initialized".format(self.Name),1)

            #store objects and change groups
            self.Objects = {}
            self.ChangeGroups = {}

            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.ConnectionMethods = ConnectionMethods(parent=self,User=self.User,Password=self.Password)
            self.StatusMethods = StatusMethods(parent=self)
        else:
            init_logger.log(lineno()+"<qsys.classes.Core class> __init__ () | {}".format(error),5)
    
    def start(self):
        if(self.connect()):
            threading.Thread(target=self.listen).start()
            threading.Thread(target=self.keepalive).start()

    def connect(self):
        try:
            self.sock.connect((self.ip,self.port))
            return True
        except Exception as e:
            self.logger.log(lineno()+"<qsys.classes.Core object - {}> connect() | {} | {}".format(self.Name,type(e).__name__,e.args),5)
            return False

    def listen(self):
        while True:
            try:
                rx = self.sock.recv(65534).decode('utf-8')
                self.parse(rx)
            except Exception as e:
                print(rx)
                self.logger.log(lineno()+"<qsys.classes.Core object - {}> listen() | {} | {}".format(self.Name,type(e).__name__,e.args),5)
            time.sleep(.001)

    def keepalive(self):
        while True:
            self.ConnectionMethods.NoOp()
            time.sleep(60)
            #clean up worker threads
            for t in threading.enumerate():
                if(not t.is_alive()):
                    t.join()

    #THIS CODE SUCKS
    def parse(self,payload): #need to think about this some more
        try:
            payload = json.loads(payload.replace("\0",""))
        except json.JSONDecodeError:
            return False
        except Exception as e:
            self.logger.log(lineno()+"<qsys.classes.Core object - {}> parse() | {} | {}".format(self.Name,type(e).__name__,e.args),5)
            return False

        ## THERE IS A MORE EFFICIENT WAY TO DO THIS
        try:
            result = payload['result']
        except KeyError:
            try:
                result = payload['params']['Changes']
            except KeyError:
                try:
                    error = payload['error']
                    self.logger.log(lineno()+"<qsys.classes.Core object - {}> error | code: {} | message: {}".format(self.Name,error['code'],error['message']),5)
                except KeyError:
                    try:
                        if('EngineStatus' in payload['method']):
                            self.__dict__.update(**payload['params']) #update __dict__ with EngineStatus tokens
                        else:
                            print("no idea...") #this is where group stuff could go
                            print(payload)
                            return False
                    except KeyError:
                        print("no idea...")
                        print(payload)
                        return False
        except Exception as e:
            #COME BACK TO THIS
            self.logger.log(lineno()+"<qsys.classes.Core object - {}> parse() | {} | {}".format(self.Name,type(e).__name__,e.args),5)
            return False

        #parse messages
        try:
            if(isinstance(result,list)):
                for item in result:
                    try:
                        self.Objects[item['Name']].state.update(**item)
                        if(self.Objects[item['Name']].state != self.Objects[item['Name']].last_state):
                            #print(self.Objects[item['Name']].state)
                            self.Objects[item['Name']].last_state = copy.copy(self.Objects[item['Name']].state)
                            self.Objects[item['Name']].change = True
                    except Exception as e:
                        self.logger.log(lineno()+"<qsys.classes.Core object - {}> parse() | {} | {}".format(self.Name,type(e).__name__,e.args),5)
            elif(isinstance(result,dict)):
                try:
                    self.Objects[result['Name']].state.update(**result)
                    if(self.Objects[result['Name']].state != self.Objects[result['Name']].last_state):
                        #print(self.Objects[result['Name']].state)
                        self.Objects[result['Name']].last_state = copy.copy(self.Objects[result['Name']].state)
                        self.Objects[result['Name']].change = True
                except Exception as e:
                    self.logger.log(lineno()+"<qsys.classes.Core object - {}> parse() | {} | {}".format(self.Name,type(e).__name__,e.args),5)
            else:
                self.logger.log(lineno()+"<qsys.classes.Core object - {}> parse() - result isn't a list or dictionary!".format(self.Name),5)
        except Exception as e:
            self.logger.log(lineno()+"<qsys.classes.Core object - {}> parse() | {} | {}".format(self.Name,type(e).__name__,e.args),5)

    def __adopt__(self,obj):
        try:
            if(type(obj) in [Control,ComponentControl,MixerControl,LoopPlayerControl]):
                self.Objects.update({obj.Name:obj})
                return True
            elif(type(obj) in [ChangeGroup]):
                self.ChangeGroups.update({obj.Id:obj})
                return True
            else:
                return False
        except Exception as e:
            self.logger.log(lineno()+"<qsys.classes.Core object - {}> __adopt__() | {} | {}".format(self.Name,type(e).__name__,e.args),5)
            return False

    def __repr__(self):
        return '<qsys.classes.Core class | Name: {Name} | Ip: {ip}>'.format(**self.__dict__)

#Control methods
class Base():    
    def struct(self):
        return {'jsonrpc':'2.0','method':'{method}','params':'{params}'}

    def encode(self,s,logger):
        try:
            return (json.dumps(s)+"\0").encode('utf-8')
        except Exception as e:
            logger.log(lineno()+"<qsys.classes.Base class> encode() | {} | {}".format(type(e).__name__,e.args),5)
            return False

    def send(self,sock,logger,**kwargs):
        tmp = self.struct()
        tmp.update(**kwargs)
        try:
            sock.send(self.encode(tmp,logger))
        except Exception as e:
            logger.log(lineno()+"<qsys.classes.Base class> send() | {} | {}".format(type(e).__name__,e.args),5)
            return False
        return True

    def __cast__(self): #allocate to parent Core class
        return self.parent.__adopt__(self)
        
class ConnectionMethods(Base):
    def __init__(self,**kwargs):
        required = {'parent':Core,'User':str,'Password':str}
        success,error = required_args(kwargs,required)
        if(success):
            self.__dict__.update(**kwargs)
        else:
            init_logger.log(lineno()+"<qsys.classes.ConnectionMethods class> __init__() | {}".format(error),5)

    def Logon(self):
        return self.send(self.parent.sock,self.parent.logger,method='logon',params={'User':self.User,'Password':self.Password})

    def NoOp(self):
        return self.send(self.parent.sock,self.parent.logger,method='NoOp',params={})
    
    def __repr__(self):
        return '<qsys.classes.ConnectionMethods object | Parent: {}>'.format(self.parent.name)

class StatusMethods(Base):
    def __init__(self,**kwargs):
        required = {'parent':Core}
        success,error = required_args(kwargs,required)
        if(success):
            self.__dict__.update(**kwargs)
        else:
            init_logger.log(lineno()+"<qsys.classes.ConnectionMethods class> __init__() | {}".format(error),5)
    
    def StatusGet(self,**kwargs):
        required = {'TransId':[int,float]}
        success,error = required_args(kwargs,required)
        if(success):
            return self.send(self.parent.sock,self.parent.logger,method='StatusGet',id=kwargs['TransId'],params=0)
        else:
            self.parent.logger.log(lineno()+"<qsys.classes.StatusMethods class> StatusGet() | {}".format(error),5)
            return False
    
    def __repr__(self):
        return '<qsys.classes.StatusMethods object | Parent: {}>'.format(self.parent.name)

class ChangeGroup(Base):
    def __init__(self,**kwargs):
        required = {'parent':Core,'Id':str}
        success,error = required_args(kwargs,required)
        if(success):
            self.__dict__.update(**kwargs)
            if(self.__cast__()):
                self.AutoPollState = False
                self.AutoPollRate = False
                self.init = True
            else:
                self.init = False
                init_logger.log(lineno()+"<qsys.classes.ChangeGroup object> __init__() | failed to __cast__() to parent",5)
        else:
            init_logger.log(lineno()+"<qsys.classes.ChangeGroup object> __init__() | {} | Args: {}".format(error,kwargs),5)
            self.init = False

    def AddControl(self,obj):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.AddControl',params={'Id':self.Id,'Controls':[obj.Name]})

    def AddComponentControl(self,obj):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.AddComponentControl',params={'Id':self.Id,'Component':{'Name':obj.Name,'Controls':[obj.Controls]}})

    def Remove(self,obj):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.Remove',params={'Id':self.Id,'Controls':[obj.Name]})

    def Destroy(self):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.Destroy',params={'Id':self.Id})

    def Invalidate(self):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.Invalidate',params={'Id':self.Id})

    def Clear(self):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.Clear',params={'Id':self.Id})

    def Poll(self,**kwargs):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.Poll',id=epoch(),params={'Id':self.Id})

    def AutoPoll(self,**kwargs):
        required = {'Rate':[int,float]}
        success,error = required_args(kwargs,required)
        if(success):
            self.AutoPollState = True
            self.AutoPollRate = kwargs['Rate']
            return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.AutoPoll',id=epoch(),params={'Id':self.Id,'Rate':self.AutoPollRate})
        else:
            self.parent.logger.log(lineno()+"<qsys.classes.qsys.ChangeGroup object - {}> AutoPoll() | {}".format(self.Id,error),5)

    def __repr__(self):
        return '<qsys.classes.ChangeGroup object | Parent: {} | Id: {}>'.format(self.parent.Name,self.Id)

class Control(Base):
    #self.init = initialized state of control
    def __init__(self,**kwargs):
        required = {'parent':Core,'Name':str,'ValueType':'*'}
        success,error = required_args(kwargs,required)
        if(success):
            self.__dict__.update(**kwargs)
            if(self.__cast__()):
                self.init = True
                self.state = {}
                self.last_state = {}
                self.change = False
                self.get(TransId=epoch())
            else:
                init_logger.log(lineno()+"<qsys.classes.Control object> __init__() | failed to __cast__() to parent",5)
                self.init = False
        else:
            init_logger.log(lineno()+"<qsys.classes.Control object> __init__() | {} | Args: {}".format(error,kwargs),5)
            self.init = False

    def get(self,**kwargs):
        required = {'TransId':[int,float]}
        success,error = required_args(kwargs,required)
        if(success):
            return self.send(self.parent.sock,self.parent.logger,method='Control.Get',id=kwargs['TransId'],params=[self.Name])
        else:
            self.parent.logger.log(lineno()+"<qsys.classes.Control object - {}> get() | {}".format(self.Name,error),5)
            return False

    def set(self,**kwargs):
        required = {'TransId':[int,float],'Value':self.ValueType} #optional 'Ramp' parameter - sets ramp time used to set the control
        success,error = required_args(kwargs,required)
        if(success):
            kwargs['Name'] = self.Name
            return self.send(self.parent.sock,self.parent.logger,method='Control.Set',id=kwargs['TransId'],params={k:v for k,v in kwargs.items() if k not in IgnoreKeys})
        else:
            self.parent.logger.log(lineno()+"<qsys.classes.Control object - {}> set() | {}".format(self.Name,error),5)
            return False

    def __repr__(self):
        return '<qsys.classes.Control object | Parent: {} | Name: {}>'.format(self.parent.name,self.Name)

class ComponentControl(Base): #coming soon
    def __init__(self,**kwargs):
        pass

class MixerControl(Base): #coming soon
    def __init__(self,**kwargs):
        pass

class LoopPlayerControl(Base): #coming soon
    def __init__(self,**kwargs):
        pass


qsys/helpers.py
import time
from inspect import currentframe

def scale_number(unscaled):
    to_min = 0;
    to_max = 100;
    from_min = -100;
    from_max = 20;
    return (to_max-to_min)*(unscaled-from_min)/(from_max-from_min)+to_min

def required_args(passed,required):
    for arg in required:
        if(arg not in passed):
            return False,'missing {}'.format(arg)
        elif(isinstance(required[arg],list)): #for multi type arguments pass a list of types - [int,float] or [str,dict]
            if(not type(passed[arg]) in required[arg]):
                return False,'{} type is {} - it should be {}'.format(arg,type(passed[arg]),required[arg])
        elif(required[arg] != '*' and not isinstance(passed[arg],required[arg])):
            return False,'{} type is {} - it should be {}'.format(arg,type(passed[arg]),required[arg])
    return True,''

def lineno(): #get line number for logging
    cf = currentframe()
    return str(cf.f_back.f_lineno)+" "

def epoch():
    return float(time.time())


.archive/trash.py
def old_test():
    core = Core(User='',Password='',ip='192.168.61.2')
    core.start()

    #gain = Control(parent=core,Name='gain',ValueType=int)
    
    cg = ChangeGroup(parent=core,Id='mygroup')

    #create some control objects
    for i in range(1,10):
        l = Control(parent=core,Name='Mixer6x9Output{}Label'.format(i),ValueType=str)
        cg.AddControl(l)
        m = Control(parent=core,Name='Mixer6x9Output{}Mute'.format(i),ValueType=[int,float])
        cg.AddControl(m)
        g = Control(parent=core,Name='Mixer6x9Output{}Gain'.format(i),ValueType=[int,float])
        cg.AddControl(g)
        print(l,m,g)

    time.sleep(2)

    cg.AutoPoll(Rate=.1)

    while True:
        x = str(input("Control Name: "))
        if(x):
            try:
                print(core.Objects[x].state)
            except:
                print("fail")


qsys/ui/static/js/application.js
var socket = io.connect('http://' + document.domain + ':' + location.port + '/test');
var controls = [];

class volDiv {
	constructor(name) {
		this.name=name;
		this.element = document.createElement('div');
		this.element.id = this.name;
		this.element.className = "progress";
		this.sub_element = document.createElement('div');
		this.sub_element.id = this.name+'_progressbar';
		this.sub_element.setAttribute('class', 'progress-bar progress-bar-striped progress-bar-animated');
		this.sub_element.setAttribute('role','progressbar');
		this.sub_element.setAttribute('aria-valuenow',this.value);
		this.sub_element.setAttribute('aria-valuemin','-100');
		this.sub_element.setAttribute('aria-valuemax','20');
		this.sub_element.setAttribute('style','width: 75%');
		this.element.appendChild(this.sub_element);
		//this.node = document.body.appendChild(this.element);
		//document.body.appendChild(document.createElement('br'));
	}
}

class controlObject {
	constructor(name) {
		this.name=name;
		this.value=0;
		this.ui_element = new volDiv(this.name);
		document.body.appendChild(this.ui_element.element);
		document.body.appendChild(document.createElement('br'));
		
	}

	listen() {
		let self = this;
		socket.on(this.name,function(msg) {
			self['value'] = msg.value;
			document.getElementById(self.name+"_progressbar").setAttribute('style','width: '+self.value+'%');
			console.log(self.name+'-'+self.value);
		});
	}
}

$(document).ready(function(){
	socket.on('payload',function(msg){
		for (i=0; i < msg.length; i++) {
			let tmp = new controlObject(msg[i]);
			//create ui object here!
			controls.push(tmp);
			controls[i].listen();
		}});
});


qsys.py
#!/usr/bin/python3
import json,socket,time,threading
from helpers import required_args,lineno,epoch
from logg3r import Log
import copy

init_logger = Log(log_path="./test_logs/",name='init_logger',level=1)

ErrorCodes = {
    -32700:'Parse error. Invalid JSON was received by the server.',
    -32600:'Invalid request. The JSON sent is not a valid Request object.',
    -32601:'Method not found.',
    -32602:'Invalid params.',
    -32603:'Server error.',
    2:'Invalid Page Request ID',
    3:'Bad Page Request - could not create the requested Page Request',
    4:'Missing file',
    5:'Change Groups exhausted',
    6:'Unknown change croup',
    7:'Unknown component name',
    8:'Unknown control',
    9:'Illegal mixer channel index',
    0:'Logon required'
}

Methods = {
    'Logon':None,
    'NoOp':None,
    'EngineStatus':{'params':{'State':str,'DesignName':str,'DesignCode':str,'IsRedundant':bool,'IsEmulator':bool}},
    'StatusGet':{'id':[int,float],'result':{'Platform':str,'State':str,'DesignName':str,'DesignCode':str,'IsRedundant':bool,'IsEmulator':bool,'Status':{'Code':int,'String':str}}}
}

IgnoreKeys = ['sock','logger','parent','ip','port']

#contains all method, control, etc objects
#handles parsing, etc - instantiate one per controlled QSC core
class Core():
    def __init__(self,**kwargs):
        required = {'User':str,
                    'Password':str,
                    'Name':str,
                    'ip':str}
        self.success,error = required_args(kwargs,required)
        if(self.success):
            try: 
                self.port = kwargs['port']
            except KeyError: 
                self.port = 1710 #standard port

            self.__dict__.update(**kwargs)
            self.logger = Log(log_path="./test_logs/",name='Core_{ip}'.format(**self.__dict__),level=1)
            self.logger.log("object initialized",1)

            #store objects and change groups
            self.Objects = {}
            self.ChangeGroups = {}

            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.ConnectionMethods = ConnectionMethods(parent=self,User=self.User,Password=self.Password)
            self.StatusMethods = StatusMethods(parent=self)
        else:
            init_logger.log(lineno()+"<qsys.Core class> __init__ () | {}".format(error),5)
    
    def start(self):
        if(self.connect()):
            threading.Thread(target=self.listen).start()
            threading.Thread(target=self.keepalive).start()

    def connect(self):
        try:
            self.sock.connect((self.ip,self.port))
            return True
        except Exception as e:
            self.logger.log(lineno()+"<qsys.Core class> connect() | {} | {}".format(type(e).__name__,e.args),5)
            return False

    def listen(self):
        while True:
            try:
                rx = self.sock.recv(65534).decode('utf-8')
                self.parse(rx)
            except Exception as e:
                print(rx)
                self.logger.log(lineno()+"<qsys.Core class> listen() | {} | {}".format(type(e).__name__,e.args),5)
            time.sleep(.001)

    def keepalive(self):
        while True:
            self.ConnectionMethods.NoOp()
            time.sleep(60)
            #clean up worker threads
            for t in threading.enumerate():
                if(not t.is_alive()):
                    t.join()

    #THIS CODE SUCKS
    def parse(self,payload): #need to think about this some more
        try:
            payload = json.loads(payload.replace("\0",""))
        except json.JSONDecodeError:
            return False
        except Exception as e:
            print(payload)
            self.logger.log(lineno()+"<qsys.Core class> parse() - {} | {}".format(type(e).__name__,e.args),5)
            return False

        ## THERE IS A MORE EFFICIENT WAY TO DO THIS
        try:
            result = payload['result']
        except KeyError:
            try:
                result = payload['params']['Changes']
            except KeyError:
                try:
                    error = payload['error']
                    self.logger.log(lineno()+"<qsys.Core class> error | code: {} | message: {}".format(error['code'],error['message']),5)
                except KeyError:
                    try:
                        if('EngineStatus' in payload['method']):
                            self.__dict__.update(**payload['params']) #update __dict__ with EngineStatus tokens
                        else:
                            print("no idea...") #this is where group stuff could go
                            print(payload)
                            return False
                    except KeyError:
                        print("no idea...")
                        print(payload)
                        return False
        except Exception as e:
            #self.logger.log(lineno()+"<qsys.Core class> parse() - {} | {}".format(type(e).__name__,e.args),5)
            #COME BACK TO THIS
            return False

        #parse messages
        try:
            if(isinstance(result,list)):
                for item in result:
                    try:
                        self.Objects[item['Name']].state.update(**item)
                        if(self.Objects[item['Name']].state != self.Objects[item['Name']].last_state):
                            #print(self.Objects[item['Name']].state)
                            self.Objects[item['Name']].last_state = copy.copy(self.Objects[item['Name']].state)
                    except Exception as e:
                        self.logger.log(lineno()+"<qsys.Core class> parse() - {} | {}".format(type(e).__name__,e.args),5)
            elif(isinstance(result,dict)):
                try:
                    self.Objects[result['Name']].state.update(**result)
                    if(self.Objects[result['Name']].state != self.Objects[result['Name']].last_state):
                        #print(self.Objects[result['Name']].state)
                        self.Objects[result['Name']].last_state = copy.copy(self.Objects[result['Name']].state)
                except Exception as e:
                    self.logger.log(lineno()+"<qsys.Core class> parse() - {} | {}".format(type(e).__name__,e.args),5)
            else:
                self.logger.log(lineno()+"<qsys.Core class> parse() - result isn't a list or dictionary!",5)
        except Exception as e:
            self.logger.log(lineno()+"<qsys.Core class> parse() - {} | {}".format(type(e).__name__,e.args),5)

    def __adopt__(self,obj):
        try:
            self.Objects.update({obj.Name:obj})
        except AttributeError:
            self.ChangeGroups.update({obj.Id:obj})

    def __repr__(self):
        return '<qsys.Core class | Name: {Name} | Ip: {ip}>'.format(**self.__dict__)

#Control methods
class Base():    
    def struct(self):
        return {'jsonrpc':'2.0','method':'{method}','params':'{params}'}

    def encode(self,s,logger):
        try:
            return (json.dumps(s)+"\0").encode('utf-8')
        except Exception as e:
            logger.log(lineno()+"<qsys.Base class> encode() | {} | {}".format(type(e).__name__,e.args),5)
            return False

    def send(self,sock,logger,**kwargs):
        tmp = self.struct()
        tmp.update(**kwargs)
        try:
            sock.send(self.encode(tmp,logger))
        except Exception as e:
            logger.log(lineno()+"<qsys.Base class> send() | {} | {}".format(type(e).__name__,e.args),5)
            return False
        return True

    def __cast__(self): #allocate to parent Core class
        self.parent.__adopt__(self)

class ConnectionMethods(Base):
    def __init__(self,**kwargs):
        required = {'parent':Core,'User':str,'Password':str}
        success,error = required_args(kwargs,required)
        if(success):
            self.__dict__.update(**kwargs)
        else:
            init_logger.log(lineno()+"<qsys.ConnectionMethods class> __init__() | {}".format(error),5)

    def Logon(self):
        return self.send(self.parent.sock,self.parent.logger,method='logon',params={'User':self.User,'Password':self.Password})

    def NoOp(self):
        return self.send(self.parent.sock,self.parent.logger,method='NoOp',params={})
    
    def __repr__(self):
        return '<qsys.ConnectionMethods class | Parent: {}>'.format(self.parent.name)

class StatusMethods(Base):
    def __init__(self,**kwargs):
        required = {'parent':Core}
        success,error = required_args(kwargs,required)
        if(success):
            self.__dict__.update(**kwargs)
        else:
            init_logger.log(lineno()+"<qsys.ConnectionMethods class> __init__() | {}".format(error),5)
    
    def StatusGet(self,**kwargs):
        required = {'TransId':[int,float]}
        success,error = required_args(kwargs,required)
        if(success):
            return self.send(self.parent.sock,self.parent.logger,method='StatusGet',id=kwargs['TransId'],params=0)
        else:
            self.parent.logger.log(lineno()+"<qsys.StatusMethods class> StatusGet() | {}".format(error),5)
            return False
    
    def __repr__(self):
        return '<qsys.StatusMethods class | Parent: {}>'.format(self.parent.name)

class ChangeGroup(Base):
    def __init__(self,**kwargs):
        required = {'parent':Core,'Id':str}
        success,error = required_args(kwargs,required)
        if(success):
            self.__dict__.update(**kwargs)
            self.__cast__()
            self.AutoPollState = False
            self.AutoPollRate = False
        else:
            init_logger.log(lineno()+"<qsys.ChangeGroup class> __init__() | {}".format(error),5)

    def AddControl(self,obj):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.AddControl',params={'Id':self.Id,'Controls':[obj.Name]})

    def AddComponentControl(self,obj):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.AddComponentControl',params={'Id':self.Id,'Component':{'Name':obj.Name,'Controls':[obj.Controls]}})

    def Remove(self,obj):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.Remove',params={'Id':self.Id,'Controls':[obj.Name]})

    def Destroy(self):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.Destroy',params={'Id':self.Id})

    def Invalidate(self):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.Invalidate',params={'Id':self.Id})

    def Clear(self):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.Clear',params={'Id':self.Id})

    def Poll(self,**kwargs):
        return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.Poll',id=epoch(),params={'Id':self.Id})

    def AutoPoll(self,**kwargs):
        required = {'Rate':[int,float]}
        success,error = required_args(kwargs,required)
        if(success):
            self.AutoPollState = True
            self.AutoPollRate = kwargs['Rate']
            return self.send(self.parent.sock,self.parent.logger,method='ChangeGroup.AutoPoll',id=epoch(),params={'Id':self.Id,'Rate':self.AutoPollRate})
        else:
            self.parent.logger.log(lineno()+"<qsys.qsys.ChangeGroup class> AutoPoll() | Id: {} | {}".format(self.Id,error),5)

    def __repr__(self):
        return '<qsys.ChangeGroup class | Parent: {} | Id: {}>'.format(self.parent.Name,self.Id)

class Control(Base):
    #self.init = initialized state of control
    def __init__(self,**kwargs):
        required = {'parent':Core,'Name':str,'ValueType':'*'}
        success,error = required_args(kwargs,required)
        if(success):
            self.__dict__.update(**kwargs)
            self.init = True
            self.state = {}
            self.last_state = {}
            self.__cast__()
            self.get(TransId=epoch())
        else:
            init_logger.log(lineno()+"<qsys.Control class> __init__() | {}".format(error),5)
            self.init = False

    def get(self,**kwargs):
        required = {'TransId':[int,float]}
        success,error = required_args(kwargs,required)
        if(success):
            return self.send(self.parent.sock,self.parent.logger,method='Control.Get',id=kwargs['TransId'],params=[self.Name])
        else:
            self.parent.logger.log(lineno()+"<qsys.Control class> set() - {} | {}".format(self.Name,error),5)
            return False

    def set(self,**kwargs):
        required = {'TransId':[int,float],'Value':self.ValueType} #option 'Ramp' parameter - sets ramp time used to set the control
        success,error = required_args(kwargs,required)
        if(success):
            kwargs['Name'] = self.Name
            #comprehension to remove TransID
            params = {k:v for k,v in kwargs.items() if k != 'TransId'}
            return self.send(self.parent.sock,self.parent.logger,method='Control.Set',id=kwargs['TransId'],params={k:v for k,v in kwargs.items() if k not in IgnoreKeys})
        else:
            self.parent.logger.log(lineno()+"<qsys.Control class> set() - {} | {}".format(self.Name,error),5)
            return False

    def __repr__(self):
        return '<qsys.Control class | Parent: {} | Name: {}>'.format(self.parent.name,self.Name)

class ComponentControl(Base): #coming soon
    def __init__(self,**kwargs):
        pass

class MixerControl(Base): #coming soon
    def __init__(self,**kwargs):
        pass

class LoopPlayerControl(Base): #coming soon
    def __init__(self,**kwargs):
        pass

if __name__ == '__main__':
    from test import main
    main()